src/wallet.js

```diff
@@ -778,23 +833,67 @@ const wallet = {
     }
 
     const isCorrect = this.isPinCorrect(oldPin);
+
     if (!isCorrect) {
       return false;
     }
 
-    // Get new PIN hash
-    const newHash = this.hashPassword(newPin);
-    // Update new PIN data in storage
-    const accessData = this.getWalletAccessData();
-    accessData['hash'] = newHash.key.toString();
-    accessData['salt'] = newHash.salt;
+    const oldAccessData = this.getWalletAccessData();
+    const newAccessData = this._getChangePinAccessData(oldAccessData, oldPin, newPin);
 
-    // Get and update data encrypted with PIN
-    const decryptedData = this.decryptData(accessData.mainKey, oldPin);
-    const encryptedData = this.encryptData(decryptedData, newPin);
-    accessData['mainKey'] = encryptedData.encrypted.toString();
+    this.setWalletAccessData(newAccessData);
 
-    this.setWalletAccessData(accessData);
+    return true;
+  },
+
+  /**
+   * Validate old password and pin and change them for the new ones
+   *
+   * This method exists so we can be sure that the `setWalletAccessData` will
+   * be called at the same time, so we can guarantee that we are not changing
+   * the PIN if the password was not changed and vice versa.
+   *
+   * This is only needed on the mobile wallet that uses the PIN as the wallet's
+   * password and needs to update them simultaneously
+   *
+   * @param {string} oldPin
+   * @param {string} newPin
+   * @param {string} oldPassword
+   * @param {string} newPassword
+   *
+   * @return {boolean} true if PIN and password were successfully changed
+   *
+   * @memberof Wallet
+   * @inner
+   */
+  changePinAndPassword(oldPin, newPin, oldPassword, newPassword) {
```

Comment:
I don't think this method should be in the lib. Changing PIN and password at the same time is something really unique for the mobile app and if the idea is to prevent success when changing pin and error when changing password, we can add `if (isPinCorrect(oldPin) && isPasswordCorrect(oldPassword)) do things` before changing both.

I'm not against the code, I think the PR is fine, so if you think it makes sense, we can go with it, I just think this shouldn't be in the lib.

If we move with this in the lib I think we should refactor this method a bit to call `changePin` and `changePassword`, so we don't repeat all the newAccessData code in the end.